qiskit/transpiler/passes/scheduling/padding.py

```diff
@@ -0,0 +1,218 @@
+    The padding pass expects all ``DAGOpNode`` in the circuit to be scheduled.
+
+    This base class doesn't define any sequence to interleave, but it manages
+    the location where the sequence is inserted, and provides a set of information necessary
+    to construct the proper sequence. Thus, a subclass of this pass just need to implement
+    :meth:`_pad` method, in which the subclass constructs a circuit block to insert.
+    This mechanism removes lots of boilerplate logic to manage whole DAG circuits.
+
+    Note that padding pass subclasses should define interleaving sequences satisfying:
+
+        - Interleaved sequence does not change start time of other nodes
+        - Interleaved sequence should have total duration of the provided ``time_interval``.
+
+    Any manipulation violating these constraints may prevent this base pass from correctly
+    tracking the start time of each instruction,
+    which may result in violation of hardware alignment constraints.
+    """
+
+    def run(self, dag: DAGCircuit):
+        """Run the padding pass on ``dag``.
+
+        Args:
+            dag (DAGCircuit): DAG to be checked.
+
+        Returns:
+            DAGCircuit: DAG with idle time filled with instructions.
+
+        Raises:
+            TranspilerError: If the whole circuit or instruction is not scheduled.
+        """
+        if "node_start_time" not in self.property_set:
+            raise TranspilerError(
+                f"The input circuit {dag.name} is not scheduled. Call one of scheduling passes "
+                f"before running the {self.__class__.__name__} pass."
+            )
+        node_start_time = self.property_set["node_start_time"]
+
+        new_dag = DAGCircuit()
+
+        for qreg in dag.qregs.values():
+            new_dag.add_qreg(qreg)
+        for creg in dag.cregs.values():
+            new_dag.add_creg(creg)
+
+        new_dag.name = dag.name
+        new_dag.metadata = dag.metadata
+        new_dag.unit = self.property_set["time_unit"]
+        new_dag.calibrations = dag.calibrations
+
+        idle_after = {bit: 0 for bit in dag.qubits}
+        circuit_duration = 0
+        for node in dag.topological_op_nodes():
+            if node in node_start_time:
+                t0 = node_start_time[node]
+                t1 = t0 + node.op.duration
+
+                if isinstance(node.op, Delay):
+                    # The padding class considers a delay instruction as idle time
+                    # rather than instruction. Delay node is removed so that
+                    # we can extract non-delay predecessors.
+                    next_node = next(dag.successors(node))
+                    if isinstance(next_node, DAGOutNode):
+                        # If user intentionally insert delay at the very end of circuit,
+                        # this should override maximum cirucit duration.
+                        circuit_duration = max(circuit_duration, t1)
```

Comment:
```suggestion
                t1 = t0 + node.op.duration
                circuit_duration = max(circuit_duration, t1)

                if isinstance(node.op, Delay):
                    # The padding class considers a delay instruction as idle time
                    # rather than instruction. Delay node is removed so that
                    # we can extract non-delay predecessors.
```
Why don't we compute `circuit_duration` from `t1`s (instead of `idle_after`)? It would be easier to read (and might be faster).